_posts/2025-04-28-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,403 @@
+
+quarkus.http.auth.permission.authenticated.paths=/mcp/sse <4>
+quarkus.http.auth.permission.authenticated.policy=authenticated
+----
+<1> Default Quarkus OIDC configuration requires that only GitHub access tokens can be used to access MCP SSE server.
+<2> By default, `quarkus.oidc.provider=github` supports an authorization code flow only. `quarkus.oidc.application-type=service` overrides it and requires the use of bearer tokens. 
+<3> Use GitHub authorization code flow to support the login endpoint with a dedicated Quarkus OIDC `login` https://quarkus.io/guides/security-openid-connect-multitenancy[tenant] configuration.
+<4> Enforce an authenticated access to the main MCP SSE endpoint during the initial handshake. See also how the tool is secured with an annotation in the <<tool>> section above.
+
+[NOTE]
+====
+Note the use of the `%prod.` prefixes. It ensures the configuration properties prefixed with `%prod.` are only effective in prod mode and not interfering with dev mode. 
+====
+
+=== Install and run application
+
+First, add `quarkus.package.jar.type=uber-jar` to the application.properties.
+
+Now, the application can be packaged and installed using:
+
+[source,shell]
+----
+mvn install
+----
+
+Run it with `jbang`:
+
+[source,shell]
+----
+jbang org.acme:secure-mcp-sse-server:1.0.0-SNAPSHOT:runner
+----
+
+### Login to GitHub and copy the access token
+
+Access `http://localhost:8080/login`, login to GitHub, and copy the returned access token:
+
+image::github_access_token.png[GitHub access token,align="center"]
+
+[[mcp-inspector]]
+=== Use MCP Inspector to access MCP server
+
+Launch https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector]:
+
+[source,shell]
+----
+npx @modelcontextprotocol/inspector
+----
+
+Paste the copied GitHub access token to the `Bearer Token` field and connect to the Quarkus MCP SSE server:
+
+image::mcp_inspector_connect.png[MCP Inspector Connect,align="center"]
+
+Next, make a `user-name-provider` tool call:
+
+image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]
+
+=== Use curl to access MCP server
+
+Finally, let's use `curl` and also learn a little bit how both the MCP protocol and MCP SSE transport work. 
+
+First, access the main SSE endpoint without the GitHub access token:
+
+[source,shell]
+----
+curl -v localhost:8080/mcp/sse
+----
+
+You will get HTTP 401 error.
+
+Use the access token to access MCP server:
+
+```shell script
+curl -v -H "Authorization: Bearer gho_..." localhost:8080/mcp/sse
+```
+
+and get an SSE response such as:
+
+[source,shell]
+----
+< content-type: text/event-stream
+< 
+event: endpoint
+data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+The SSE connection is created.
+
+Now open another window and use the same access token to initialize the curl as MCP client, and access the tool, using the value of the `data` property to build the target URL.
```

Comment:
Add something like:

We cannot invoke the tool directly, we first need to follow the MCP handshake protocol: